kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/storage/PartitionLog.java

```diff
@@ -526,6 +537,14 @@ public CompletableFuture<ReadRecordsResult> readRecords(final FetchRequest.Parti
                 requestStats.getPrepareMetadataStats().registerSuccessfulEvent(
                         MathUtils.elapsedNanos(startPrepareMetadataNanos), TimeUnit.NANOSECONDS);
                 long adjustedMaxBytes = Math.min(partitionData.maxBytes, limitBytes.get());
+                if (readCommitted && producerStateManager.firstUndecidedOffset().isPresent()
+                        && producerStateManager.firstUndecidedOffset().get().compareTo(offset) <= 0) {
+                    long highWaterMark = MessageMetadataUtils.getHighWatermark(cursor.getManagedLedger());
+                    future.complete(ReadRecordsResult.empty(highWaterMark,
+                            this.firstUndecidedOffset().orElse(highWaterMark),
```

Comment:
In this case, `firstUndecidedOffset()` should always return a non-empty optional and its value should be the same with `producerStateManager.firstUndecidedOffset().get()`. How about fixing the logic like:

```java
                if (readCommitted) {
                    long firstUndecidedOffset = producerStateManager.firstUndecidedOffset().orElse(-1L);
                    if (firstUndecidedOffset >= 0 && firstUndecidedOffset <= offset) {
                        long highWaterMark = MessageMetadataUtils.getHighWatermark(cursor.getManagedLedger());
                        future.complete(ReadRecordsResult.empty(highWaterMark, firstUndecidedOffset,
                                tcm.getManagedLedger().getLastConfirmedEntry()));
                        return;
                    }
                }
```

Reply:
fixed